pkg/cdi/filter.go

```diff
@@ -0,0 +1,160 @@
+package cdi
+
+import (
+	glob "path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Filter can be used for matching vendors, classes or devices.
+// Filters can use both verbatim strings and glob-style patterns.
+type Filter struct {
```

Comment:
I was considering something similar for a short while. But TBH, I simply couldn't find a single use case to justify overdoing it that much (which I guess is slightly surprising), not even in the cdi test utility.

The only real use cases for filtering IMO are filtering for a single vendor (cdi test utility vendors command), for a single particular class (cdi test utility classes command), or for a particular class of a particular vendor (Kevin's "I need to inject all of my vendor's GPUs into a single container").

Actually after implementing Filter as an exported type, I started regretting it/having second thoughts about it. Currently I think that the cleanest approach would be to let (potential) globbing be 'explicitly implied' by the function name itself providing *Matching*() API calls (ListMatchingDevices(), InjectMatchingDevices(), etc.) and have these functions internally construct/handle filtering without an explicit Filter being exposed to the user at all.

Reply:
After thinking about this some more, I think that we should drop the filtering entirely for *this iteration of the API* and keep the implementation of the `*Matching*` API calls as follow ups. I added a general comment to #38 to this regards.

I also agree that hiding these from the user is a cleaner approach.

With regards to the utility, this can still be achieved by post-processing a returned list of `Spec`s, correct, so maybe move the filtering to the utility for the time being if we think that we need something more than "list everything" for the time being.

Reply:
Sorry, I failed to notice #38 as I was busy trying to address those comments that I both fully understood and agreed with.  I pushed my updates before having noticed your further comments. So just for the record, I did the hiding you mention up there in the latest PR update, by switching Filter to filter, passing string patterns instead of filters through the public API and constructing a filter from a pattern on the API ingress. But I agree. Let's remove it for now and reconsider it later once we have managed to agree on some basic functionality and got it merged.

The only reason I was keen to implement the *Matching* interfaces was to check if it would be a good fit to trivially implement the 'inject all devices of a particular class from vendorA' use case without having to define in the CDI spec an 'all' device with something like this:
```
func InjectFullVendorDeviceClass(spec *oci.Spec, vendor, class string) error {
    pattern := vendor+"/"+class+"=*"
    matching, err := cdi.MatchDevices(vendor+"/"+class+"=*")
    if err != nil {
        return errors.Wrapf(err, "failed to collect devices matching pattern %q", pattern)
    }
    return cdi.InjectDevices(matching)
}
``` 
I tested it and it works fine, so I'm OK with dropping it and reconsidering later.